bl/routefunctions.go

```
@@ -0,0 +1,129 @@
+
+func DeleteBook(c *gin.Context){
+	username    := c.DefaultQuery("username","no_value")
+	id := c.Param("id")
+	client := dal.ElasticInitClient()
+	response := dal.ElasticDeleteBook(client, id)
+	updateUserData(username,"Delete " +c.Request.RequestURI)
+	c.JSON(http.StatusOK,response)
+}
+
+func SearchBook(c *gin.Context){
+	username    := c.DefaultQuery("username","no_value")
+	title       := c.DefaultQuery("title","no_value")
+	authorName  := c.DefaultQuery("author_name","no_value")
+	priceRangeS := c.DefaultQuery("price_range", "no_value")
+	client      := dal.ElasticInitClient()
+	price_range := getPriceRange(priceRangeS)
+	var opcode int = 0
+	if fieldExist(priceRangeS){
+		opcode +=1
+	}
+	if fieldExist(title){
+		opcode +=2
+	}
+	if fieldExist(authorName){
+		opcode +=4
+	}
+	searchResult := dal.ElasticSearchBook(client, title, authorName, price_range, opcode)
+	updateUserData(username,"GET " +c.Request.RequestURI)
+	// Iterate through results
+	for _, hit := range searchResult.Hits.Hits {
+		c.JSON(http.StatusOK, hit.Source)
+	}
+}
+
+func GetStoreData(c *gin.Context){
+	username    := c.DefaultQuery("username","no_value")
+	client := dal.ElasticInitClient()
+	count,diffAuthors := dal.ElasticGetStoreData(client)
+	updateUserData(username,"GET " +c.Request.RequestURI)
+	c.JSON(http.StatusOK, count)
+	c.JSON(http.StatusOK, diffAuthors.Hits.TotalHits)
+}
+func GetUserData(c *gin.Context){
+	username    := c.DefaultQuery("username","no_value")
+	if username == "no_value"{
+		panic("No Username Given")
+	}
+	userData := dal.ReddisGetUserRecord(username)
+	c.JSON(http.StatusOK, userData)
+}
+
+func fieldExist(check string) (bool){
+	if check == "no_value"{
+		return false
+	}
+	return true
+}
+
+func getPriceRange(priceRangeS string) ([2]float64){
+	priceRange := [2]float64{0.0,0.0}
+	var err error
+	if !fieldExist(priceRangeS){
+		return priceRange
+	}
+	priceRange[0],err = strconv.ParseFloat(strings.Split(priceRangeS,"-")[0], 64)
+	if err!=nil{
+		panic(err)
```

you should return the error and let the caller to decided how to handle it